### ISSUE TITLE
cli crash when using history

### ISSUE THREAD

**Issue by justinc1** [cloudius-systems/osv#716]
enter say ls /asdf, then cursor up, then cursor down. It crashes with backtrace. Seems to be well (always?) reproducible.

**Comment by justinc1** [comment]
A hopefully useful gdb where
```
(gdb) where
#0  processor::cli_hlt () at arch/x64/processor.hh:248
#1  0x0000000000209697 in arch::halt_no_interrupts () at arch/x64/arch.hh:48
#2  0x000000000049b2eb in osv::halt () at arch/x64/power.cc:24
#3  0x000000000022bc94 in abort (fmt=0x99ebcc "Aborted\n") at runtime.cc:132
#4  0x000000000022bb74 in abort () at runtime.cc:96
#5  0x0000000000687722 in osv::generate_signal (siginfo=..., ef=0xffff800005226078) at libc/signal.cc:128
#6  0x0000000000687801 in osv::handle_mmap_fault (addr=86802432, sig=11, ef=0xffff800005226078) at libc/signal.cc:139
#7  0x00000000003c661b in mmu::vm_sigsegv (addr=86802432, ef=0xffff800005226078) at core/mmu.cc:1317
#8  0x00000000003c67a3 in mmu::vm_fault (addr=86802432, ef=0xffff800005226078) at core/mmu.cc:1337
#9  0x000000000048bae6 in page_fault (ef=0xffff800005226078) at arch/x64/mmu.cc:38
#10 <signal handler called>
#11 0x0000100005c156f6 in tputs ()
#12 0x000010000581c411 in terminal_tputs ()
#13 0x000010000581a733 in terminal_move_to_char ()
#14 0x0000100005816d90 in re_update_line ()
#15 0x000010000581627a in re_refresh ()
#16 0x0000100005815a63 in el_gets ()
#17 0x0000100005401b7b in main (argc=<optimized out>, argv=<optimized out>) at cli.c:222
#18 0x000000000062663c in osv::application::run_main (this=0xffffa000047d1118, path="/cli/cli.so", argc=1, argv=0xffffa00002f5cc00) at core/app.cc:293
#19 0x000000000062686d in osv::application::run_main (this=0xffffa000047d1118) at core/app.cc:310
#20 0x0000000000626340 in osv::application::main (this=0xffffa000047d1118) at core/app.cc:238
#21 0x0000000000625f4e in osv::application::__lambda5::operator() (__closure=0x0, app=0xffffa000047d1118) at core/app.cc:188
#22 0x0000000000625f73 in osv::application::__lambda5::_FUN (app=0xffffa000047d1118) at core/app.cc:190
#23 0x0000000000681eed in pthread_private::pthread::pthread(void* (*)(void*), void*, sigset_t, pthread_private::thread_attr const*)::{lambda()#1}::operator()() const ()
    at libc/pthread.cc:101
#24 0x0000000000684524 in std::_Function_handler<void(), pthread_private::pthread::pthread(void* (*)(void*), void*, sigset_t, const pthread_private::thread_attr*)::__lambda6>::_M_invoke(const std::_Any_data &) (__functor=...) at /usr/include/c++/4.8/functional:2071
#25 0x000000000044c688 in std::function<void ()>::operator()() const (this=0xffff800005221080) at /usr/include/c++/4.8/functional:2471
#26 0x00000000005bc8aa in sched::thread::main (this=0xffff800005221050) at core/sched.cc:1051
#27 0x00000000005b9045 in sched::thread_main_c (t=0xffff800005221050) at arch/x64/arch-switch.hh:164
#28 0x000000000048b8f6 in thread_main () at arch/x64/entry.S:113

```

**Comment by nyh** [comment]
Doesn't happen on my system! cursor up brings back the "ls /asdf" line, and then a cursor down brings an empty line again. Pressing cursor up or down too many times (beyond the end) causes a beep. No crash...
I wonder what's the difference on your machine... I'm using gcc 5.3.1 on Fedora 22, what are you using?

In any case, according to your stack trace, the crash is coming from deep inside the "libedit" library which cli.c is using (calling its el_gets() function). So it might be that we have nothing to fix except upgrading the libedit library that we use and hope for the best....

**Comment by nyh** [comment]
I've confirmed that this bug does not exist when compiling OSv on Fedora 22 or 23, both with gcc 5.3.1.
@justinc1's buggy image was compiled on Ubuntu 14.04.03 LTS with gcc 4.8.4-2 and boost 1.54.0.1.

He sent me the image he compiled, and it does have the bug reported here. Moreover, this image  has an even more obvious bug in the CLI: Pressing backspace (\177 or ^?, actually) produces 7 characters of garbage to the screen instead of erasing the last character as happens on the image I compile. The backspace does its function (cancel one previous inputted character) but outputs garbage to the screen. Moreover, this only happens in the CLI's libedit line editor - if I run "top" in the CLI (to go out of the CLI's line editor and into OSv's standard cooked line discipline) the backspace works normally.

All of this might be a bug in the libedit library. Would be nice (for someone who can reproduce these CLI bugs) to try to update modules/libedit to pick up a newer version of this library - maybe it really had a bug which was already fixed.

**Comment by nyh** [comment]
I compiled OSv myself on an Ubuntu 14.04.03 fully updated (same gcc version as Justin reported), and indeed, I can reproduce exactly the same problems reported above in the CLI (backspace outputs crap, certain lineup-linedown combinations crash, etc.).

**Comment by nyh** [comment]
Ok, I found the problem. It is really embarassing build bug in libedit. It's not a general OSv bug, and only effects the CLI.

TL;DR: try "apt-get install ncurses-dev" and then compile OSv from scratch (scripts/build clean; scripts/build image=cli). Your CLI woes should be gone :-) Please let me know if you can reproduce that.

When libedit is compiled without the "curses" library installed on the host, it, quite embarassingly, doesn't "#include <curses.h>", but continues does use terminfo functions like tgetent() which no other header files defines... The compiler warns loudly that a function with no prototype has been used, that integers are converted to pointers, and other crap like that, but the user is happy to ignore all these messages and use the broken result :-)

If you install the ncurses-dev, libedit's "configure" will find that curses.h exists, will "#include" it, and the warnings will be gone - along with the broken behavior.

Halleluja!

Now to think how to make sure this doesn't happen to anyone in the future... Possible directions to consider:
 1. Get libedit's configure to use ncurses from ../module/ncurses, not the one installed on the host. 
 2. Install ncurses-dev on the host in scripts/setup.py
 3. Report this problem upstream so they fix libedit's "configure".

**Comment by justinc1** [comment]
Nadav, thank you for all the work.
